types/styled-components/index.d.ts

```diff
@@ -10,6 +10,10 @@
 //                 Aaron Reisman <https://github.com/lifeiscontent>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
+// This is needed to make other libs happy when TS is importing this file
+// while testing others.
+/// <reference lib="dom" />
```

Comment:
I'm not sure packages should be allowed to do this. It's adding an implicit environment guarantee that doesn't necessarily exist (styled-components could be in a react-native environment for example).

Instead, do something like the `global.d.ts` file in the react types and add an empty declaration for `interface ShadowRoot {}` inside the `declare global` block.

Reply:
Thanks for reopening and for the pointers.

I have been running `styled-components` inside a ShadowRoot like this for a while. But always had to add a `@ts-expect-error` to pass the type-checking.

Is there anything specific that is not supported when running in a ShadowRoot?

Reply:
It seems I just misunderstood a detail about styled-components' implementation (`adoptedStyleSheets` is a CSSOM API and exists on both `Document` and `ShadowRoot`, but styled-components' "CCSOM" implementation seems to use an older API that depends on style tags instead of constructed sheets). This seems to matter on `closed`-mode roots.